### PR TITLE
Custom Token Check Exemption Rules

### PR DESCRIPTION
**Pull request by JVickery-TBS**
feat(dev): added custom exemption rules;

- Added config option to add custom exemption rules from token checks.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -91,6 +91,10 @@ Optional
     # Default 10 minutes.
     ckanext.csrf_filter.token_rotation_minutes = 10
 
+    # Exempts given regex matches from token checks.
```
Comment (ThrawnCA) [inline README.md:94]: Most sites will want multiple rules, so the necessary syntax should be described here.

File: ckanext/csrf_filter/anti_csrf.py [hunk 2]
```diff
@@ -90,6 +91,9 @@ def configure(config):
     token_expiry_age = 60 * config.get('ckanext.csrf_filter.token_expiry_minutes', 30)
     token_renewal_age = 60 * config.get('ckanext.csrf_filter.token_renewal_minutes', 10)
 
+    exempt_rules = config.get('ckanext.csrf_filter.exempt_rules', None)
+    if exempt_rules:
+        exempt_rules = [re.compile(rule) for rule in exempt_rules.split()]
```
Comment (ThrawnCA) [inline ckanext/csrf_filter/anti_csrf.py:96]: So, splitting by whitespace?

File: ckanext/csrf_filter/anti_csrf.py [hunk 3]
```diff
@@ -90,6 +91,9 @@ def configure(config):
     token_expiry_age = 60 * config.get('ckanext.csrf_filter.token_expiry_minutes', 30)
     token_renewal_age = 60 * config.get('ckanext.csrf_filter.token_renewal_minutes', 10)
 
+    exempt_rules = config.get('ckanext.csrf_filter.exempt_rules', None)
```
Comment (ThrawnCA) [inline ckanext/csrf_filter/anti_csrf.py:94]: `API_URL` should be subsumed into this list.

**Review by ThrawnCA — CHANGES_REQUESTED** [review]
Looks good overall, just some areas for improvement. Thanks for submitting this.

File: ckanext/csrf_filter/anti_csrf.py [hunk 4]
```diff
@@ -90,6 +91,9 @@ def configure(config):
...
+    if exempt_rules:
```
Comment (duttonw) [inline ckanext/csrf_filter/anti_csrf.py:95]: Seems like a good option. This does warrant an actual unit test which would also show examples as something this 'powerful' left without proper unit testing/examples can/will shoot people in the foot for overly lax regex rules etc.

File: ckanext/csrf_filter/anti_csrf.py (hunk 2 above)
Comment (JVickery-TBS) [reply ckanext/csrf_filter/anti_csrf.py]: Changed it to be a JSON list. Seems better and is easier to read an write in INI files. Added this info to the readme for the config option

File: README.md (hunk 1 above)
Comment (JVickery-TBS) [reply README.md:99]: Added required syntax to readme

File: ckanext/csrf_filter/anti_csrf.py (hunk 3 above)
Comment (JVickery-TBS) [reply ckanext/csrf_filter/anti_csrf.py]: `API_URL` value is now the default exempt rule. The custom ones from the config value will be added on if they exist

File: ckanext/csrf_filter/anti_csrf.py (hunk 4 above)
Comment (JVickery-TBS) [reply ckanext/csrf_filter/anti_csrf.py]: A warning has been added to the readme for this config option telling users to make sure they do strict regex statements.

I have added in a new test which shows a use case with the Datatables view plugin. Which was my reasoning for doing exempt rules to begin with as the Datatables stuff happens in JS and not HTML so no token fields get injected to it

File: ckanext/csrf_filter/test_anti_csrf.py [hunk 5]
```diff
@@ -248,6 +249,21 @@ def _check_config(self, secret_key, secure_cookies=False,
         self.assertEqual(anti_csrf.token_expiry_age, token_expiry_age)
         self.assertEqual(anti_csrf.token_renewal_age, token_renewal_age)
 
+    def test_exempt_rules(self):
+        """ Tests that the custom exempt rules pass their conditions
```
Comment (ThrawnCA) [inline ckanext/csrf_filter/test_anti_csrf.py:253]: This is not very descriptive. Perhaps "Test that requests matching the exemption rules are not checked for tokens" or something similar.

**Review by ThrawnCA — APPROVED** [review]

File: ckanext/csrf_filter/anti_csrf.py [hunk 6]
```diff
@@ -95,6 +96,12 @@ def configure(config):
     token_expiry_age = 60 * config.get('ckanext.csrf_filter.token_expiry_minutes', 30)
     token_renewal_age = 60 * config.get('ckanext.csrf_filter.token_renewal_minutes', 10)
 
+    exempt_rules = [re.compile(r'^/+api/.*')]
+    custom_exempt_rules = config.get('ckanext.csrf_filter.exempt_rules', None)
+    if custom_exempt_rules:
+        for rule in json.loads(custom_exempt_rules):
```
Comment (duttonw) [inline ckanext/csrf_filter/anti_csrf.py:102]: can you add a test for a invalid json load.

File: ckanext/csrf_filter/test_anti_csrf.py [hunk 7]
```diff
@@ -70,7 +70,8 @@ def mock_objects(username=None):
     """
     anti_csrf.configure({
         'ckanext.csrf_filter.secret_key': 'secret',
-        'ckan.site_url': 'https://unit-test'})
+        'ckan.site_url': 'https://unit-test',
+        'ckanext.csrf_filter.exempt_rules': '["^/datatables/ajax/.*"]'})
```
Comment (duttonw) [inline ckanext/csrf_filter/test_anti_csrf.py:74]: can you add another test with multiple rules and checking that they are loaded correctly into the exempt_rules list.

File: ckanext/csrf_filter/anti_csrf.py (hunk 6 above)
Comment (JVickery-TBS) [reply ckanext/csrf_filter/anti_csrf.py:102]: Test for invalid json has been added

**Comment by JVickery-TBS** [comment]
@duttonw test scenarios from above have been added now

**Review by ThrawnCA — APPROVED** [review]

**Pull request merged**